router/broker.go

```diff
@@ -898,3 +1008,119 @@ func (b *broker) subCountSubscribers(msg *wamp.Invocation) wamp.Message {
 		Arguments: wamp.List{count},
 	}
 }
+
+// syncGetSubHistoryEvents returns all stored events for subscription
+// TODO: Need to filter by authid and/or authrole of caller if stored events have any of them
+// But that's not possible in current arch as meta peer doesn't know anything about caller, just invocation message
+func (b *broker) syncGetSubHistoryEvents(subId wamp.ID) []*storedEvent {
+	if subscription, ok := b.subscriptions[subId]; ok {
+		if storeItem, ok := b.eventHistoryStore[subscription]; ok {
+			var events []*storedEvent
+			for _, event := range storeItem.events {
+				events = append(events, event.event)
+			}
+			return events
+		}
+	}
+
+	return []*storedEvent{}
+}
+
+// eventHistoryLast retrieves N last events for subscription.
+func (b *broker) eventHistoryLast(msg *wamp.Invocation) wamp.Message {
+	var events wamp.List // []*storedEvent
+	subId, ok1 := wamp.AsID(msg.Arguments[0])
+	limit, ok2 := msg.Arguments[1].(int)
+
+	if !ok1 || !ok2 || limit <= 0 {
+		return &wamp.Error{
+			Type:    msg.MessageType(),
+			Request: msg.Request,
+			Details: wamp.Dict{},
+			Error:   wamp.ErrInvalidArgument,
+		}
+	}
+
+	sync := make(chan struct{})
+	b.actionChan <- func() {
+		storedEvents := b.syncGetSubHistoryEvents(subId)
+		filteredEvents := storedEvents[-limit:]
```

Comment:
Wait, negative? Am I missing something here?

Reply:
Nope) That is me missing something. I was doing that offline (and thus can not even get docs) and did not remember how start:end indexes work in go. So a bit of a mess from go/py/js worlds %) 
Fixed!